File: celladapter/src/main/java/io/erva/celladapter/CellAdapter.java
package io.erva.celladapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.util.ArrayMap;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CellAdapter extends RecyclerView.Adapter<Cell> {

    private final Map<Class, Class<? extends Cell>> itemCellMap = new ArrayMap<>();
    private final List<Class> viewTypes = new ArrayList<>();
    private final SparseArray<Cell.Listener<?>> typeListenerMapping = new SparseArray<>();
    private List items = new ArrayList<>();

    public void registerCell(@NonNull Class<?> itemClass,
                             @NonNull Class<? extends Cell> cellClass) {
        registerCell(itemClass, cellClass, null);
    }

    public void registerCell(@NonNull Class<?> itemClass,
                             @NonNull Class<? extends Cell> cellClass,
                             @Nullable Cell.Listener<?> cellListener) {
        itemCellMap.put(itemClass, cellClass);
        int type = viewTypes.indexOf(itemClass);
        if (type == -1) viewTypes.add(itemClass);
        if (cellListener != null) registerListener(itemClass, cellListener);
    }

    private void registerListener(Class<?> itemClass,
                                  @Nullable Cell.Listener<?> cellListener) {
        int index = viewTypes.indexOf(itemClass);
        if (index < 0)
            throw new IllegalStateException(itemClass.getSimpleName() + " is not registered as Cell");
        typeListenerMapping.put(index, cellListener);
    }

    @Override
    public Cell onCreateViewHolder(ViewGroup parent, int viewType) {
        Class<? extends Cell> cellClass = itemCellMap.get(viewTypes.get(viewType));
        Cell cell = buildCell(cellClass, parent);
        cell.setCellDelegate(typeListenerMapping.get(viewType));
        return cell;
    }

    private Cell buildCell(Class<? extends Cell> cellClass, ViewGroup parent) {
        Layout layoutAnnotation = cellClass.getAnnotation(Layout.class);
        View cellView = LayoutInflater.from(parent.getContext()).inflate(layoutAnnotation.value(), parent, false);
        RecyclerView.ViewHolder cellObject = null;
        try {
            Constructor<? extends RecyclerView.ViewHolder> constructor = cellClass.getConstructor(View.class);
            cellObject = constructor.newInstance(cellView);
        } catch (Exception e) {
            Log.e("CellAdapter", "Can't create cell: " + e.getMessage());
        }
        return (Cell) cellObject;
    }

    @Override
    public void onBindViewHolder(Cell cell, int position) {
        cell.bindViewInternal(getItem(position));
    }

    @Override
    public int getItemViewType(int position) {
        Class itemClass = items.get(position).getClass();
        int index = viewTypes.indexOf(itemClass);
        if (index < 0) {
            throw new IllegalArgumentException(itemClass.getSimpleName() + " is not registered");
        }
        return index;
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public boolean contains(Object item) {
        return items.contains(item);
    }

    public int indexOf(Object item) {
        return items.indexOf(item);
    }

    public void replaceItem(int position, Object item) {
        items.set(position, item);
    }

    public Object getItem(int position) {
        return items.get(position);
    }

    public List getItems() {
        return items;
    }

    public void setItems(List items) {
        this.items = items;
    }

    public void addItem(int position, Object item) {
        items.add(position, item);
    }

    public void addItems(List items) {
        if (items != null) {
            this.items.addAll(items);
        }
    }

    public void remove(Object item) {
        if (item != null) {
            int position = indexOf(item);
            if (position != -1) remove(position);
        }
    }

    public void remove(int position) {
        items.remove(position);
    }

    public void clear() {
        items.clear();
    }
}

File: celladapter-kotlin/src/main/kotlin/io/erva/celladapter/v7/CellAdapter.kt
package io.erva.celladapter.v7

import android.support.v7.widget.RecyclerView
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.ViewGroup
import io.erva.celladapter.Layout
import java.util.*
import kotlin.reflect.KClass

open class CellAdapter : RecyclerView.Adapter<Cell<out Any, Cell.Listener<out Any>>>() {

    private val itemCellMapping = HashMap<KClass<*>, KClass<out Cell<out Any, Cell.Listener<out Any>>>>()
    private val viewTypes: MutableList<KClass<*>> = mutableListOf()
    private val itemListenerMapping = SparseArray<Cell.Listener<out Any>>()

    val items: MutableList<Any> = mutableListOf()

    open fun cell(cellClass: KClass<out Cell<out Any, Cell.Listener<out Any>>>,
                  cellDataBuilderFunc: CellDataBuilder.() -> CellDataBuilder): CellAdapter {
        val cellDataBuilder = CellDataBuilder().cellDataBuilderFunc()
        itemCellMapping.put(cellDataBuilder.itemClass, cellClass)
        val type = viewTypes.indexOf(cellDataBuilder.itemClass)
        if (type == -1) {
            this.viewTypes.add(cellDataBuilder.itemClass)
            if (cellDataBuilder.listener != null)
                itemListenerMapping.put(viewTypes.indexOf(cellDataBuilder.itemClass), cellDataBuilder.listener)
        }
        return this
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Cell<out Any, Cell.Listener<out Any>> {
        val cellClass = itemCellMapping[viewTypes[viewType]]!!
        val layoutAnnotation = cellClass.annotations
                .filter { it is Layout }
                .map { it as Layout }
                .firstOrNull() ?: throw IllegalStateException("You have to specify Layout annotation")

        val cellView = LayoutInflater.from(parent.context).inflate(layoutAnnotation.layout, parent, false)
        return cellClass.constructors.first().call(cellView)
    }

    override fun getItemViewType(position: Int): Int {
        val baseItemClass = this.items[position]
        val index = viewTypes.indexOf(baseItemClass.javaClass.kotlin)
        if (index < 0) {
            throw IllegalArgumentException(baseItemClass.javaClass.simpleName + " is not registered")
        }
        return index
    }

    override fun onBindViewHolder(cell: Cell<out Any, Cell.Listener<out Any>>, position: Int) {
        val item = getItem(position)
        cell.listener(itemListenerMapping.get(getItemViewType(position)))
        cell.item(item)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    operator fun contains(item: Any): Boolean {
        return items.contains(item)
    }

    fun indexOf(item: Any): Int {
        return items.indexOf(item)
    }

    fun replaceItem(position: Int, item: Any) {
        items[position] = item
    }

    fun getItem(position: Int): Any {
        return items[position]
    }

    fun addItem(item: Any) {
        items.add(item)
    }

    fun addItem(position: Int, item: Any) {
        items.add(position, item)
    }

    fun addItems(items: List<Any>?) {
        if (items != null) {
            this.items.addAll(items)
        }
    }

    fun remove(item: Any?) {
        items.remove(item)
    }

    fun remove(position: Int) {
        items.removeAt(position)
    }

    fun clear() {
        items.clear()
    }

    class CellDataBuilder() {

        lateinit var itemClass: KClass<*>
        var listener: Cell.Listener<out Any>? = null

        fun item(itemClass: KClass<*>): CellDataBuilder {
            this.itemClass = itemClass
            return this
        }

        fun listener(listener: Cell.Listener<out Any>): CellDataBuilder {
            this.listener = listener
            return this
        }
    }
}